halo2_proofs/src/dev/cost.rs

```diff
@@ -38,34 +39,151 @@ pub struct CircuitCost<G: PrimeGroup, ConcreteCircuit: Circuit<G::Scalar>> {
     permutation_cols: usize,
     /// Number of distinct sets of points in the multiopening argument.
     point_sets: usize,
+    /// Maximum rows used over all columns
+    max_rows: usize,
+    /// Maximum rows used over all advice columns
+    max_advice_rows: usize,
+    /// Maximum rows used over all fixed columns
+    max_fixed_rows: usize,
+    num_fixed_columns: usize,
+    num_advice_columns: usize,
+    num_instance_columns: usize,
+    num_total_columns: usize,
 
     _marker: PhantomData<(G, ConcreteCircuit)>,
 }
 
-struct Assembly {
-    selectors: Vec<Vec<bool>>,
+/// Region implementation used by Layout
+#[derive(Debug)]
+pub(crate) struct LayoutRegion {
+    /// The name of the region. Not required to be unique.
+    pub(crate) name: String,
+    /// The columns used by this region.
+    pub(crate) columns: HashSet<RegionColumn>,
+    /// The row that this region starts on, if known.
+    pub(crate) offset: Option<usize>,
+    /// The number of rows that this region takes up.
+    pub(crate) rows: usize,
+    /// The cells assigned in this region. We store this as a `Vec` so that if any cells
+    /// are double-assigned, they will be visibly darker.
+    pub(crate) cells: Vec<(RegionColumn, usize)>,
+}
+
+/// Cost and graphing layouter
+#[derive(Default, Debug)]
+pub(crate) struct Layout {
+    /// k = 1 << n
+    pub(crate) k: u32,
+    /// Regions of the layout
+    pub(crate) regions: Vec<LayoutRegion>,
+    current_region: Option<usize>,
+    /// Total row count
+    pub(crate) total_rows: usize,
+    /// Total advice rows
+    pub(crate) total_advice_rows: usize,
+    /// Total fixed rows
+    pub(crate) total_fixed_rows: usize,
+    /// Any cells assigned outside of a region. We store this as a `Vec` so that if any
+    /// cells are double-assigned, they will be visibly darker.
+    pub(crate) loose_cells: Vec<(RegionColumn, usize)>,
+    /// Pairs of cells between which we have equality constraints.
+    pub(crate) equality: Vec<(Column<Any>, usize, Column<Any>, usize)>,
+    /// Selector assignments used for optimization pass
+    pub(crate) selectors: Vec<Vec<bool>>,
+}
+
+impl Layout {
+    /// Creates a empty layout
+    pub fn new(k: u32, n: usize, num_selectors: usize) -> Self {
+        Layout {
+            k,
+            regions: vec![],
+            current_region: None,
+            total_rows: 0,
+            total_advice_rows: 0,
+            total_fixed_rows: 0,
+            /// Any cells assigned outside of a region. We store this as a `Vec` so that if any
+            /// cells are double-assigned, they will be visibly darker.
```

Comment:
These comments don't need to be duplicated here.

Reply:
Fixed!